p2p/starknet/handlers.go

```diff
@@ -43,133 +39,130 @@ func getBuffer() *bytes.Buffer {
 		return
 	}
 
-	var req spec.Request
-	if err := proto.Unmarshal(buffer.Bytes(), &req); err != nil {
-		h.log.Debugw("Error unmarshalling message", "peer", stream.ID(), "protocol", stream.Protocol(), "err", err)
+	var zero ReqT
+	req := zero.ProtoReflect().New().Interface()
+	if err := proto.Unmarshal(buffer.Bytes(), req); err != nil {
+		log.Debugw("Error unmarshalling message", "peer", stream.ID(), "protocol", stream.Protocol(), "err", err)
 		return
 	}
 
-	response, err := h.reqHandler(&req)
+	response, err := reqHandler(req.(ReqT))
 	if err != nil {
-		h.log.Debugw("Error handling request", "peer", stream.ID(), "protocol", stream.Protocol(), "err", err, "request", req.String())
+		log.Debugw("Error handling request", "peer", stream.ID(), "protocol", stream.Protocol(), "err", err)
 		return
 	}
 
 	for msg, valid := response(); valid; msg, valid = response() {
 		if _, err := protodelim.MarshalTo(stream, msg); err != nil { // todo: figure out if we need buffered io here
-			h.log.Debugw("Error writing response", "peer", stream.ID(), "protocol", stream.Protocol(), "err", err)
+			log.Debugw("Error writing response", "peer", stream.ID(), "protocol", stream.Protocol(), "err", err)
 		}
 	}
 }
 
-func (h *Handler) reqHandler(req *spec.Request) (Stream[proto.Message], error) {
-	var singleResponse proto.Message
-	var err error
-	switch typedReq := req.GetReq().(type) {
-	case *spec.Request_GetBlocks:
-		return h.HandleGetBlocks(typedReq.GetBlocks)
-	case *spec.Request_GetSignatures:
-		singleResponse, err = h.HandleGetSignatures(typedReq.GetSignatures)
-	case *spec.Request_GetEvents:
-		singleResponse, err = h.HandleGetEvents(typedReq.GetEvents)
-	case *spec.Request_GetReceipts:
-		singleResponse, err = h.HandleGetReceipts(typedReq.GetReceipts)
-	case *spec.Request_GetTransactions:
-		singleResponse, err = h.HandleGetTransactions(typedReq.GetTransactions)
-	default:
-		return nil, fmt.Errorf("unhandled request %T", typedReq)
-	}
+func (h *Handler) BlockHeadersHandler(stream network.Stream) {
+	streamHandler[*spec.BlockHeadersRequest](stream, h.OnBlockHeadersRequest, h.log)
+}
 
-	if err != nil {
-		return nil, err
-	}
-	return StaticStream[proto.Message](singleResponse), nil
+func (h *Handler) BlockBodiesHandler(stream network.Stream) {
+	streamHandler[*spec.BlockBodiesRequest](stream, h.OnBlockBodiesRequest, h.log)
+}
+
+func (h *Handler) EventsHandler(stream network.Stream) {
+	streamHandler[*spec.EventsRequest](stream, h.OnEventsRequest, h.log)
 }
 
-func (h *Handler) HandleGetBlocks(req *spec.GetBlocks) (Stream[proto.Message], error) {
+func (h *Handler) ReceiptsHandler(stream network.Stream) {
+	streamHandler[*spec.ReceiptsRequest](stream, h.OnReceiptsRequest, h.log)
+}
+
+func (h *Handler) TransactionsHandler(stream network.Stream) {
+	streamHandler[*spec.TransactionsRequest](stream, h.OnTransactionsRequest, h.log)
+}
+
+func (h *Handler) OnBlockHeadersRequest(req *spec.BlockHeadersRequest) (Stream[proto.Message], error) {
```

Comment:
Can we un-export the `On...Request` methods?

Reply:
done, might need to export them again at some point for testing